### PR TITLE
Pending Token Transfers

### PR DESCRIPTION
**Pull request by Nana-EC**
**Description**:
Users on the Hedera network have the ability to control the token classes that can be added to their balance. A user must first associate with a token class before they may receive token value from others. This differs from other networks where users may receive token value with no restrictions - in short, airdrops are currently not supported on the Hedera network. For some, this serves as a fundamental account protection feature, for others, this serves as an extra user step required before receiving desired token. 

[HIP 655](https://github.com/hashgraph/hedera-improvement-proposal/pull/655) provides a suggestion for a token outbox concept that tracks a sending users intent to airdrop tokens to a receiving account (EOA or contract) that has not yet associated with the token class.

This outbox gives a mapping that may be used to enable the simulation of the airdrop transfer concept that web3 users are familiar with. A network can communicate a senders commitment to transfer a token to a potential recipient, the intended recipient maintains account sovereignty as the network enforces the association requirement before a receiver takes ownership of the tokens. Tokens in an outbox are conceptually pending token transfers or offers to the desired recipient until they accept or reject the intent. Upon acceptance the intended token value will be added to the recipients balance. Additionally, support for association and transfer of tokens in an atomic operation will allow the seamless ability to take ownership and transfer from your pending balance.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by bmgentile** [comment]
I believe this is going to negatively impact decentralized exchanges on the Hedera network; here's how:

**1. Liquidity Pool Dynamics:**
• Delayed Finality: If someone swaps for a certain token type on a DEX but doesn’t immediately associate the token, the actual transfer of tokens from the liquidity pool is delayed. This would mean the liquidity pool’s total assets and the ratio of tokens within it would not reflect real-time transactions, creating a lag in the reflection of actual liquidity.
• Potential for Sudden Drains: If multiple users decide to associate their tokens simultaneously after a delay, it could lead to sudden and substantial drains from the liquidity pool. This could destabilize the pool, especially if it’s not very large.

**2. Price Impact and Slippage:**
Swapping tokens on DEXs often comes with price impacts, especially for large trades or low-liquidity tokens. If the actual liquidity pool does not reflect pending associations, it could mislead users about the potential price impact of their swaps. When the tokens are eventually associated and deducted, the sudden change in liquidity might result in unexpected slippage for subsequent traders.

**3. Arbitrage Opportunities:**
This delayed token association mechanism might be exploited by arbitrageurs. They could monitor the pending associations and act on price discrepancies between the DEX and other exchanges before the liquidity pool adjusts.

**4. User Experience and Complexity:**
• For the end-users, this could introduce another layer of complexity. They’d need to be aware of the association mechanism and its implications on their trades. Some might forget to associate, leading to confusion or concerns about missing funds.
• Users might also use the association delay tactically, waiting for favorable market conditions before associating and finalizing their trades. This could lead to speculative behavior.

**5. Operational Overheads for DEX:**
The DEX would need mechanisms to monitor and handle pending associations. This could increase the operational complexity and resource usage.

**Comment by bmgentile** [comment]
If the goal is to solve for airdrops specifically, I think there would need to be a specific transaction type called “airdrop” or the like, which performs this activity.

I believe that would be an awesome and **very** desired useful feature by application developers.

But if it’s all-encompassing with any token transfer, it starts to cause problems across the ecosystem, like the above example.

**Comment by Nana-EC** [comment]
> If the goal is to solve for airdrops specifically, there could be a transaction type on Hedera called “Airdrop Transfer” and that specific transaction type allows someone to send tokens with this "pending" feature. This way it would silo the airdrop ability and be an AMAZING feature for our users (and one that's frequently requested) / support growing the adoption of Hedera.
> 
> It would hurt many applications if it’s not a very specific transaction type to enable "pending", and all transfers of tokens had it enabled.

@bmgentile HIP 655's goal is to fix airdrops. This HIPs goal is to ease the friction of onboarding and that first experience of transferring tokens to an account that's not associated. Airdrops update a recipients balance without them taking action. This allows them to take action to approve or reject, so this is more airgrab-ish

**Comment by Nana-EC** [comment]
> If the goal is to solve for airdrops specifically, there could be a transaction type on Hedera called “Airdrop Transfer” and that specific transaction type allows someone to send tokens with this "pending" feature. This way it would silo the airdrop ability and be an AMAZING feature for our users (and one that's frequently requested) / support growing the adoption of Hedera.
> 
> It would hurt many applications if it’s not a very specific transaction type to enable "pending", and all transfers of tokens had it enabled.

@bmgentile All transfer that are applicable to this state would utilize the outbox. Should a user have to take action to receive the token after the transfer why would this hurt applications?
Can you highlight some scenarios to help weigh where there are still gaps in the solution.

**Comment by Nana-EC** [comment]
> If the goal is to solve for airdrops specifically, there could be a transaction type on Hedera called “Airdrop Transfer” and that specific transaction type allows someone to send tokens with this "pending" feature. This way it would silo the airdrop ability and be an AMAZING feature for our users (and one that's frequently requested) / support growing the adoption of Hedera.
> 
> It would hurt many applications if it’s not a very specific transaction type to enable "pending", and all transfers of tokens had it enabled.

@bmgentile A new transaction type just to circumvent token associate wouldn't work as it'd break the sovereignty of an account choice on the network to be associated or not with a token. Users should be able to chose the behaviour their account takes (use auto association etc) for HTS as it differs from ERC tokens in these advanced management features.

**Comment by Nana-EC** [comment]
> I believe this is going to negatively impact decentralized exchanges on the Hedera network; here's how:
> 
> **1. Liquidity Pool Dynamics:** • Delayed Finality: If someone swaps for a certain token type on a DEX but doesn’t immediately associate the token, the actual transfer of tokens from the liquidity pool is delayed. This would mean the liquidity pool’s total assets and the ratio of tokens within it would not reflect real-time transactions, creating a lag in the reflection of actual liquidity. • Potential for Sudden Drains: If multiple users decide to associate their tokens simultaneously after a delay, it could lead to sudden and substantial drains from the liquidity pool. This could destabilize the pool, especially if it’s not very large.
> 
> **2. Price Impact and Slippage:** Swapping tokens on DEXs often comes with price impacts, especially for large trades or low-liquidity tokens. If the actual liquidity pool does not reflect pending associations, it could mislead users about the potential price impact of their swaps. When the tokens are eventually associated and deducted, the sudden change in liquidity might result in unexpected slippage for subsequent traders.
> 
> **3. Arbitrage Opportunities:** This delayed token association mechanism might be exploited by arbitrageurs. They could monitor the pending associations and act on price discrepancies between the DEX and other exchanges before the liquidity pool adjusts.
> 
> **4. User Experience and Complexity:** • For the end-users, this could introduce another layer of complexity. They’d need to be aware of the association mechanism and its implications on their trades. Some might forget to associate, leading to confusion or concerns about missing funds. • Users might also use the association delay tactically, waiting for favorable market conditions before associating and finalizing their trades. This could lead to speculative behavior.
> 
> **5. Operational Overheads for DEX:** The DEX would need mechanisms to monitor and handle pending associations. This could increase the operational complexity and resource usage.

@bmgentile these are solid illustrations, thanks.
The DeFi complexity of liquidity management on top of this feature does indeed make it challenging.
I'm not sure how the system can both debit a sender and credit a potential recipient without the potential recipients first associating when it comes to HTS tokens.
Hedera DApps can manage this challenge by checking for association and enabling users, however, EVM native DApps aren't yet willing to add this additional Hedera account feature set to their  flow as it's not an EVM native feature.

Indeed complexity is increased.
I will continue to think on this. 🤔

File: HIP/hip-0000-pending-token-transfers.md [hunk 1]
```diff
@@ -1,165 +1,201 @@
...
 
-[HIP 655](https://github.com/hashgraph/hedera-improvement-proposal/pull/655) provides a suggestion for a token outbox concept that tracks a sending users intent to airdrop tokens to a receiving account (EOA or contract) that has not yet associated with the token class.
+[HIP 655](https://github.com/hashgraph/hedera-improvement-proposal/pull/655) suggested the token outbox concept to track a sending users intent to airdrop tokens to a receiving account (EOA or contract) that has not yet associated with the token class. It also noted an inbox which would hold non-assocaited tokens in a separate location from an accounts main balance.
```
Comment (dougvk) [inline HIP/hip-0000-pending-token-transfers.md:19]: associate

File: HIP/hip-0000-pending-token-transfers.md [hunk 2]
```diff
@@ -1,165 +1,201 @@
...
 
-Notably, the Hedera network offers an established paradigm to transparently and securely associate with HTS tokens prior to receiving token value. This ensures sovereignty of accounts and allows them to take explicit actions of token ownership which may have financial and legal consequences. 
+Notably, the Hedera network offers an established paradigm to transparently and securely associate with HTS tokens prior to receiving token value. This ensures sovereignty of accounts and allows them to take explicit actions of token ownership which may have financial and legal consequences. Account also offer auto assocaition slots to open their balance up to recieve non-associated tokens. However, achieving this state requires additional steps by users adn DApps and requires they have already onboarded and are knowledgeable of Hedera features on top of EVM features
```
Comment (dougvk) [inline HIP/hip-0000-pending-token-transfers.md:27]: association

File: HIP/hip-0000-pending-token-transfers.md [hunk 3]
```diff
@@ -337,11 +394,19 @@ The reference implementation must be complete before any HIP is given the status
 3. Remote token assocaiton limitation ont he network with the abilityt to opt-in for restricted association
```
Comment (dougvk) [inline HIP/hip-0000-pending-token-transfers.md:394]: association

**Comment by dougvk** [comment]
@bmgentile @Nana-EC I'm not so sure this is a problem for DEXs. One of the classic user flows for depositing liquidity into a pool or staking into a pool is having to approve this or that token before you can even interact with the pool itself. Approvals to me sound the same as associations. In fact, they're more strict -- sometimes they are infinite approvals -- but Ethereum hacks have shown that that's a bad practice. So usually you have to go through approvals every time you want to deposit or withdraw. Jumping through these sorts of hoops is normal in Ethereum DEX UX flows.

**Review by mgarbs — APPROVED** [review]

**Pull request merged**